theta/include/compact_theta_sketch_parser_impl.hpp

```diff
@@ -29,32 +29,97 @@ template<bool dummy>
+          if (num_entries == 0) {
+              return {true, true, seed_hash, 0, theta_constants::MAX_THETA, nullptr};
+          } else {
+              const size_t expected_size_bytes = (preamble_size + num_entries) << 3;
+              if (size < expected_size_bytes) {
+                  throw std::invalid_argument(std::to_string(expected_size_bytes) + " bytes expected, actual " + std::to_string(size)
+                      + (dump_on_error ? (", sketch dump: " + hex_dump(reinterpret_cast<const uint8_t*>(ptr), size)) : ""));
+              }
+              const uint64_t* entries = reinterpret_cast<const uint64_t*>(ptr) + COMPACT_SKETCH_ENTRIES_EXACT_U64;
+              return {false, true, seed_hash, num_entries, theta_constants::MAX_THETA, entries};
+          }
+      } else if (preamble_size == 3) {
+          const uint32_t num_entries = reinterpret_cast<const uint32_t*>(ptr)[COMPACT_SKETCH_NUM_ENTRIES_U32];
+          uint64_t theta = reinterpret_cast<const uint64_t*>(ptr)[COMPACT_SKETCH_THETA_U64];
+          bool is_empty = (num_entries == 0) && (theta == theta_constants::MAX_THETA);
+          if (is_empty) {
+              return {true, true, seed_hash, 0, theta, nullptr};
+          }
+          const uint64_t* entries = reinterpret_cast<const uint64_t*>(ptr) + COMPACT_SKETCH_ENTRIES_ESTIMATION_U64;
+          const size_t expected_size_bytes = (COMPACT_SKETCH_ENTRIES_ESTIMATION_U64 + num_entries) * sizeof(uint64_t);
+          if (size < expected_size_bytes) {
+            throw std::invalid_argument(std::to_string(expected_size_bytes) + " bytes expected, actual " + std::to_string(size)
+                + (dump_on_error ? (", sketch dump: " + hex_dump(reinterpret_cast<const uint8_t*>(ptr), size)) : ""));
+          }
+          return {false, false, seed_hash, num_entries, theta, entries};
+      } else {
+          throw std::invalid_argument(std::to_string(preamble_size) + " longs of premable, but expected 1, 2, or 3");
+      }
   }
-  const uint32_t num_entries = reinterpret_cast<const uint32_t*>(ptr)[COMPACT_SKETCH_NUM_ENTRIES_U32];
-  const size_t entries_start_u64 = has_theta ? COMPACT_SKETCH_ENTRIES_ESTIMATION_U64 : COMPACT_SKETCH_ENTRIES_EXACT_U64;
-  const uint64_t* entries = reinterpret_cast<const uint64_t*>(ptr) + entries_start_u64;
-  const size_t expected_size_bytes = (entries_start_u64 + num_entries) * sizeof(uint64_t);
-  if (size < expected_size_bytes) {
-    throw std::invalid_argument(std::to_string(expected_size_bytes) + " bytes expected, actual " + std::to_string(size)
-        + (dump_on_error ? (", sketch dump: " + hex_dump(reinterpret_cast<const uint8_t*>(ptr), size)) : ""));
+  default:
+      // this should always fail since the valid cases are handled above
+      checker<true>::check_serial_version(reinterpret_cast<const uint8_t*>(ptr)[COMPACT_SKETCH_SERIAL_VERSION_BYTE], COMPACT_SKETCH_SERIAL_VERSION);
```

Comment:
I left the check in place because it generates the better error message. Unfortunately, the compiler couldn't detect that the default case would always throw an exception in that case, causing a warning that one path (the default case) did not return a value. I added the throw, which should never be reached, to get rid of the warning.

Reply:
I don't quite like this, but not a problem. Perhaps can be addressed later. This also reminds me that I wanted to call this parser from the regular deserialization to avoid code duplication, but never got around to it.